Build lib llvm dynamic lib

winksaville opened the pull request:
@SeanTAllen see if this fixes the need to delete the system provided llvm.

mfelsche commented:
Appveyor is good.

winksaville commented:
This is ready to merge.

SeanTAllen commented:
@winksaville should we have a CI job that tests this? Seems like a good idea to me, something generic, so we make sure it doesn't get broken. Thoughts?

If yes, what would the make command we want to run for that look like?

winksaville commented:
I think I might have a logic error, hold off merging.

winksaville commented:
When the CI checks finish, this is again ready to merge.

There was **no logic error**, the problem was actually in my `support-simultaneous-llvm-checkouts` branch. But there was a comment error, `if we're statically linking` is now changed to `if we're dynamically linking` and added a couple pedantic comments lines :
```
   # Set rpath for ponyc if we're statically linking LLVM_VENDOR
  ifeq (,$(filter $(use), llvm_link_static))
    ifeq ($(LLVM_VENDOR),true)
      llvm.ldflags += -Wl,-rpath,$(llvm.libdir)
    endif
  endif
```
So it now reads:
```
  # Set rpath for ponyc if we're dynamically linking LLVM_VENDOR
  ifeq (,$(filter $(use), llvm_link_static))
    # We're linking dynamically as "llvm_link_static" is empty.
    ifeq ($(LLVM_VENDOR),true)
      # LLVM_VENDOR is true, so have linker set rpath.
      llvm.ldflags += -Wl,-rpath,$(llvm.libdir)
    endif
  endif
```

winksaville commented:
So maxOS-release failed, it looks like it is testing with LLVM 8.0.0 as I see the following in the logs:
```
/usr/local/Cellar/llvm/8.0.0/include/llvm/IR/DIBuilder.h
```
It seems they just upgraded to 8.0.0, https://formulae.brew.sh/formula/llvm. On the website they don't have a page for 7.0.1 :( I'll changed `.cirrus.yml and see what happens.

winksaville commented:
@SeanTAllen, txs for the fix, this is ready for review and merging!

winksaville commented:
@SeanTAllen @mfelsche, could we get a review on this and maybe merge so I'm not tempted to make more changes :)

mfelsche commented:
@winksaville i would love to review this, but i don't have enough context to know what this should fix and what the actual effect is. If you could add this into the PR description, that would help me a lot. Then i can also test it locally.

srenatus commented on Makefile-ponyc:
```diff
@@ -291,7 +291,20 @@ ifneq ($(MAKECMDGOALS),clean)
+    endif
+  endif
+
+  $(warning llvm.libdir="$(llvm.libdir)")
```
I'm missing some context, and warning might be the only "log level" available. If there's a way to make this something like "info" or "debug", it could be preferable.

winksaville replied:
Txs, there is a $(info) I wasn't aware of, I'll change to that.

srenatus commented on Makefile-lib-llvm:
```diff
@@ -33,7 +33,7 @@ pony_targets += stdlib test-stdlib stdlib-debug test-stdlib-debug test-examples
 .PHONY: $(pony_targets)
 $(pony_targets): $(llvm_config)
-	@PATH=$(new_path) $(MAKE) -f Makefile-ponyc config=$(config) $(MAKECMDGOALS) $(MAKEFLAGS)
+	@PATH=$(new_path) $(MAKE) -f Makefile-ponyc LLVM_VENDOR=true config=$(config) $(MAKECMDGOALS) $(MAKEFLAGS)
```
[nit] `PATH` is set before `$(MAKE)`, I suppose we could set the new var there, too. OTOH,  `PATH` might affect `$(MAKE)`, so, there's a reason to treat it differently.

winksaville replied:
[$(MAKE)](https://www.gnu.org/software/make/manual/html_node/MAKE-Variable.html) is "magical" in gnu make so using it is unrelated to setting PATH. I set PATH so that when projects besides llvm are requested to be built by there presence in `llvm_project_list` such as clang and clang++ they are found before any system provided programs of the same name.

One thing I will do is expicitly set LLVM_CONFIG where appropriate. Setting PATH does mean `llvm-config` is found by Makefile-ponyc, but its better to be explicit.

srenatus commented on lib/llvm/Makefile:
```diff
@@ -97,13 +97,8 @@ LLVM_INSTALL_DIR := $(ROOT_DIR)/dist
-#else
-#  LLVM_USE_LINKER := -DLLVM_USE_LINKER=$(LLVM_LINKER)
-#endif
+#LLVM_LINKER := bfd
```
Let's add a little note why you might want `bfd`, or remove the comment?

winksaville replied:
Good idea

SeanTAllen commented:
Do not merge until after #3121 has been merged.

EpicEric commented:
#3121 has been merged

winksaville commented:
Rebased onto master, worked for me on my computer, if it turns green then ready for last minute reviews and merging, fingers crossed, merging.
